service/controller/v1/resource/drainer/create.go

```diff
@@ -0,0 +1,159 @@
+	// a strategic merge patch in order to drain them.
+	UnschedulablePatch = `{"spec":{"unschedulable":true}}`
+)
+
+// EnsureCreated represents the node resource implementation to manage on demand
+// node draining for guest clusters.
+func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
+	drainerConfig, err := key.ToDrainerConfig(obj)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	if drainerConfig.Status.HasFinalCondition() {
+		r.logger.LogCtx(ctx, "level", "debug", "message", "node config status already has final state")
+		resourcecanceledcontext.SetCanceled(ctx)
+		r.logger.LogCtx(ctx, "level", "debug", "message", "canceling resource for custom object")
+
+		return nil
+	}
+
+	var k8sClient kubernetes.Interface
+	{
+		i := key.ClusterIDFromDrainerConfig(drainerConfig)
+		e := key.ClusterEndpointFromDrainerConfig(drainerConfig)
+		k8sClient, err = r.guestCluster.NewK8sClient(ctx, i, e)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	}
+
+	{
+		r.logger.LogCtx(ctx, "level", "debug", "message", "cordoning guest cluster node")
+
+		n := key.NodeNameFromDrainerConfig(drainerConfig)
+		t := types.StrategicMergePatchType
+		p := []byte(UnschedulablePatch)
+
+		_, err := k8sClient.CoreV1().Nodes().Patch(n, t, p)
+		if apierrors.IsNotFound(err) {
+			// It might happen the node we want to drain got already removed. This
+			// might even be due to human intervention. In case we cannot find the
+			// node we assume the draining was successful and set the node config
+			// status accordingly.
+
+			r.logger.LogCtx(ctx, "level", "debug", "message", "guest cluster node not found")
+			r.logger.LogCtx(ctx, "level", "debug", "message", "setting node config status of guest cluster node to final state")
+
+			drainerConfig.Status.Conditions = append(drainerConfig.Status.Conditions, drainerConfig.Status.NewFinalCondition())
+
+			_, err := r.g8sClient.CoreV1alpha1().DrainerConfigs(drainerConfig.GetNamespace()).UpdateStatus(&drainerConfig)
+			if err != nil {
+				return microerror.Mask(err)
+			}
+
+			r.logger.LogCtx(ctx, "level", "debug", "message", "set node config status of guest cluster node to final state")
+			resourcecanceledcontext.SetCanceled(ctx)
+			r.logger.LogCtx(ctx, "level", "debug", "message", "canceling resource for custom object")
+
+			return nil
+		} else if err != nil {
+			return microerror.Mask(err)
+		}
+
+		r.logger.LogCtx(ctx, "level", "debug", "message", "cordoned guest cluster node")
+	}
+
+	var customPods []v1.Pod
+	var systemPods []v1.Pod
+	{
+		r.logger.LogCtx(ctx, "level", "debug", "message", "looking for all pods running on the guest cluster node")
+
+		fieldSelector := fields.SelectorFromSet(fields.Set{
+			"spec.nodeName": key.NodeNameFromDrainerConfig(drainerConfig),
+		})
+		listOptions := apismetav1.ListOptions{
+			FieldSelector: fieldSelector.String(),
+		}
+		podList, err := k8sClient.CoreV1().Pods(v1.NamespaceAll).List(listOptions)
```

Comment:
I think it would be a good idea to have paging support here. Technically it's not guaranteed what is a size of a single result page and on larger clusters it could be that single call to `List()` doesn't return all results.

Reply:
Same here, nothing changed with this. Once there is a problem we can adapt. I also never heard about paging with the k8s client.

Reply:
Ah, you are right. It seems to be currently in development.